File: app/lib/web/models/event.dart
// Mirrors api/Models/Event.cs. One row per thing that has ever happened to the caller -
// a bird departing/arriving, a hub post, a friend request being accepted - written once and
// never pruned server-side. The journey log reads the full list; the notification dropdown
// reads the isNotification subset.
class AppEvent {
  final String id;
  final String kind;
  final String displayText;
  final String? quotedNote;
  final String? targetType;
  final String? targetId;
  final bool isNotification;
  final bool isRead;
  final DateTime createdAt;
  // Null for anything that isn't "someone else did this to you" (e.g. your own bird departing
  // or arriving) - where set (a friend's bird arriving at your nest, a friend accepting your
  // request), this is that friend's userId, letting the web UI tint the notification with
  // their own friend color instead of a generic unread tint.
  final String? sourceUserId;

  AppEvent({
    required this.id,
    required this.kind,
    required this.displayText,
    this.quotedNote,
    this.targetType,
    this.targetId,
    required this.isNotification,
    required this.isRead,
    required this.createdAt,
    this.sourceUserId,
  });

  factory AppEvent.fromJson(Map<String, dynamic> json) => AppEvent(
    id: json['id'] as String,
    kind: json['kind'] as String,
    displayText: json['displayText'] as String,
    quotedNote: json['quotedNote'] as String?,
    targetType: json['targetType'] as String?,
    targetId: json['targetId'] as String?,
    isNotification: json['isNotification'] as bool,
    isRead: json['isRead'] as bool,
    createdAt: DateTime.parse(json['createdAt'] as String),
    sourceUserId: json['sourceUserId'] as String?,
  );
}

// Matches api/Models/Event.cs's EventKind constants.
class EventKind {
  static const birdDeparted = 'BirdDeparted';
  static const birdArrived = 'BirdArrived';
  static const birdArrivedAtYourNest = 'BirdArrivedAtYourNest';
  static const hubPostCreated = 'HubPostCreated';
  static const birdJoinedFlock = 'BirdJoinedFlock';
  static const friendAdded = 'FriendAdded';
  static const friendRequestAccepted = 'FriendRequestAccepted';
  static const friendRequestReceived = 'FriendRequestReceived';
}

// Matches api/Models/Event.cs's EventTargetType constants.
class EventTargetType {
  static const bird = 'Bird';
  static const nest = 'Nest';
  static const hub = 'Hub';
}

File: api/Models/Event.cs
using System.Text.Json.Serialization;

namespace CroApp.Api.Models;

// Id is a generated GUID, and UserId (whose timeline/notification list this belongs to) is
// the partition key - same single-partition-per-owner reasoning as Waypoint/Bird. Written
// once, at the moment something happens (a bird departs/arrives, a hub post lands, a friend
// request is accepted), and never updated afterward except IsRead flipping via the
// notifications endpoints. DisplayText is fully rendered at write time from whatever data is
// available then - never re-derived from a live Bird/HubMessage/User lookup later, because
// those can be deleted (Bird) or TTL-expire (HubMessage) while this record must not. This is
// the one place in the backend that intentionally never gets pruned - the web UI's "journey
// log" reads this as a permanent history, and the notification bell reads the IsNotification
// subset of it. QuotedNote/TargetType/TargetId are optional: QuotedNote carries a snapshotted
// payload excerpt for the timeline's quoted-note blocks, TargetType/TargetId let a
// notification or timeline entry deep-link back to the still-live Bird/Nest/Hub it's about
// (null once that's not applicable, e.g. a friend-request notification just routes to the
// Friends screen client-side). SourceUserId is null for anything that isn't "someone else did
// this to you" (self-only kinds like BirdDeparted/BirdArrived never set it) - where it is set
// (BirdArrivedAtYourNest, FriendRequestAccepted) it's the other user's id, letting the web UI
// tint that notification with the sender's own friend color without re-deriving it from
// DisplayText, which is free-form and not guaranteed to name them.
public record Event(
    [property: JsonPropertyName("id")] string Id,
    string UserId,
    string Kind,
    string DisplayText,
    string? QuotedNote,
    string? TargetType,
    string? TargetId,
    bool IsNotification,
    bool IsRead,
    DateTimeOffset CreatedAt,
    string? SourceUserId = null);

public static class EventKind
{
    public const string BirdDeparted = "BirdDeparted";
    public const string BirdArrived = "BirdArrived";
    public const string BirdArrivedAtYourNest = "BirdArrivedAtYourNest";
    public const string HubPostCreated = "HubPostCreated";
    public const string BirdJoinedFlock = "BirdJoinedFlock";
    public const string FriendAdded = "FriendAdded";
    public const string FriendRequestAccepted = "FriendRequestAccepted";
    public const string FriendRequestReceived = "FriendRequestReceived";
}

public static class EventTargetType
{
    public const string Bird = "Bird";
    public const string Nest = "Nest";
    public const string Hub = "Hub";
}
